media-placeholders/src/main/java/org/wordpress/aztec/placeholders/PlaceholderManager.kt

```diff
@@ -0,0 +1,338 @@
+import android.graphics.drawable.Drawable
+import android.text.Editable
+import android.text.Layout
+import android.text.Spanned
+import android.view.View
+import android.widget.FrameLayout
+import androidx.core.content.ContextCompat
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import org.wordpress.aztec.AztecAttributes
+import org.wordpress.aztec.AztecContentChangeWatcher
+import org.wordpress.aztec.AztecText
+import org.wordpress.aztec.Constants
+import org.wordpress.aztec.plugins.html2visual.IHtmlTagHandler
+import org.wordpress.aztec.spans.AztecMediaClickableSpan
+import org.xml.sax.Attributes
+import java.util.UUID
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * This class handles the "Placeholders". Placeholders are custom spans which are drawn in the Aztec text and the user
+ * can interact with them in a similar way as with other media. These spans are invisible and are used by this class
+ * as a place we can draw over with custom views. The custom views are placed in the `FrameLayout` which contains the
+ * Aztec text item and are shifted up and down if anything above them changes (for example if the user adds a new line
+ * before the placeholder).
+ */
+class PlaceholderManager(
+        private val aztecText: AztecText,
+        private val container: FrameLayout
+) : AztecContentChangeWatcher.AztecTextChangeObserver,
+        IHtmlTagHandler,
+        AztecText.OnMediaDeletedListener,
+        CoroutineScope {
+    private val job = Job()
+    override val coroutineContext: CoroutineContext
+        get() = Dispatchers.Main + job
+    private val drawers = mutableMapOf<String, PlaceholderAdapter>()
+    private val positionToId = mutableSetOf<Placeholder>()
+
+    init {
+        aztecText.contentChangeWatcher.registerObserver(this)
+    }
+
+    fun onDestroy() {
+        aztecText.contentChangeWatcher.unregisterObserver(this)
+        job.cancel()
+        drawers.clear()
+    }
+
+    /**
+     * Register a custom drawer to draw a custom view over a placeholder.
+     */
+    fun registerDrawer(placeholderAdapter: PlaceholderAdapter) {
+        drawers[placeholderAdapter.type] = placeholderAdapter
+    }
+
+    /**
+     * Call this method to manually insert a new item into the aztec text. There has to be a drawer associated with the
+     * item type.
+     * @param id placeholder ID
+     * @param type placeholder type
+     * @param attributes other attributes passed to the view. For example a `src` for an image.
+     */
+    fun insertItem(id: String, type: String, vararg attributes: Pair<String, String>) {
+        if (drawers[type] == null) throw IllegalArgumentException("Drawer for inserted type not found. Register it with `registerDrawer` method")
+        val attrs = getAttributesForMedia(id, type, attributes)
+        val drawer = drawers[type] ?: return
```

Comment:
Recommend simplifying:

```suggestion
        val drawer = drawers[type] ?: throw IllegalArgumentException("Drawer for inserted type not found. Register it with `registerDrawer` method")
        val attrs = getAttributesForMedia(id, type, attributes)
```